vigranumpy/src/core/segmentation.cxx

```diff
@@ -1188,6 +1191,70 @@ python::tuple  pyUnionFindWatershedsBlockwise(
     return python::make_tuple(out, nSeg);
 }
 
+/** \brief Map all values in src to new values using the given mapping (a dict).
+ *  See python docstring for details.
+*/
+template <unsigned int NDIM, class SrcVoxelType, class DestVoxelType>
+NumpyAnyArray
+pythonApplyMapping(NumpyArray<NDIM, Singleband<SrcVoxelType> > src,
+                   python::dict mapping,
+                   bool allow_incomplete_mapping = false,
+                   NumpyArray<NDIM, Singleband<DestVoxelType> > res = NumpyArray<NDIM, Singleband<SrcVoxelType> >())
+{
+    using namespace boost::python;
+
+    res.reshapeIfEmpty(src.taggedShape(), "applyMapping(): Output array has wrong shape.");
+
+    // Copy dict into a c++ unordered_map of ints,
+    // which is ~10x faster than using a Python dict
+    typedef boost::unordered_map<SrcVoxelType, DestVoxelType> labelmap_t;
+    labelmap_t labelmap(2*len(mapping)); // Using 2*N buckets seems to speed things up by 10%
+
+    typedef stl_input_iterator<tuple> dict_iter_t;
+
+#if PY_MAJOR_VERSION < 3
+    dict_iter_t map_iter = mapping.iteritems();
+#else
+    dict_iter_t map_iter = mapping.items();
+#endif
+
+    for (; map_iter != dict_iter_t(); ++map_iter)
+    {
+        object key = (*map_iter)[0];
+        object value = (*map_iter)[1];
+        labelmap[extract<uint32_t>(key)] = extract<uint32_t>(value);
+    }
+
+    PyAllowThreads _pythread;
```

Comment:
This must always go into a scope below `return res` because the latter manipulates Python objects and therefore needs the GIL (which is only acquired after the return in the present design).